### PR TITLE
[lock/unlock] OP L1 finalization in bridge adapter

### PR DESCRIPTION
**Pull request by makramkd**
## Motivation

The Optimism L1 bridge adapter needs to be able to both prove withdrawals and finalize them on L1.

## Solution

* Add Optimism withdrawal proving and finalization on L1 functionality to the bridge adapter. Tweak the scripts to prove and finalize withdrawals through the adapters rather than on the Optimism contracts directly to prove that they work.
* Fix a bug with ETH cross-chain transfers, on both L1 and L2 adapters: since Optimism transfers ETH rather than WETH cross-chain through their native bridge, stuff ends up reverting when we call `withdraw` on the WETH contract prior to bridging.
* Fix a bug where we were using `withdrawTo` without actually sending the ETH along with the call. The Optimism contracts dependency we're using isn't bedrock so the `withdrawTo` function signature did not have `payable`. Instead we copy/paste the `withdrawTo` function signature directly from [`L2StandardBridge.sol`](https://github.com/ethereum-optimism/optimism/blob/f707883038d527cbf1e9f8ea513fe33255deadbc/packages/contracts-bedrock/src/L2/L2StandardBridge.sol#L114-L122), since we can't import it directly due to the exact 0.8.15 version specified in the contract's pragma.

## Follow ups

* Direct follow up: support multiple-stage finalization in Rebalancer.sol. Right now Rebalancer.sol assumes that a single successful call to `finalizeWithdrawERC20` will give it access to funds, however this is not the case for Optimism. This will be implemented in an upcoming PR to keep diffs small and manageable.
* Future work: Explore the differences between `bridgeETHTo` and `bridgeERC20To`, which are on `StandardBridge.sol`, vs. `depositERC20To` and `withdrawTo` which are on L1StandardBridge.sol and L2StandardBridge.sol respectively.
* Future work: Use contracts-bedrock imports instead of plain contracts.

### CODE REVIEW

File: contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol [hunk 1]
```diff
@@ -5,33 +5,46 @@ import {IBridgeAdapter} from "../interfaces/IBridge.sol";
 import {IWrappedNative} from "../../ccip/interfaces/IWrappedNative.sol";
 
 import {IL1StandardBridge} from "@eth-optimism/contracts/L1/messaging/IL1StandardBridge.sol";
-import {IL1CrossDomainMessenger} from "@eth-optimism/contracts/L1/messaging/IL1CrossDomainMessenger.sol";
 import {IERC20} from "../../vendor/openzeppelin-solidity/v4.8.3/contracts/token/ERC20/IERC20.sol";
 import {SafeERC20} from "../../vendor/openzeppelin-solidity/v4.8.3/contracts/token/ERC20/utils/SafeERC20.sol";
 
+import {Types} from "../interfaces/optimism/Types.sol";
```
Comment (matYang) [inline contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol]: nit: import ordering https://github.com/smartcontractkit/chainlink/blob/develop/contracts/STYLE_GUIDE.md#imports

File: contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol [hunk 2]
```diff
@@ -72,41 +85,102 @@ contract OptimismL1BridgeAdapter is IBridgeAdapter {
     return 0;
   }
 
+  /// @notice Payload for proving a withdrawal from L2 on L1 via finalizeWithdrawERC20.
+  /// @param withdrawalTransaction The withdrawal transaction, see its docstring for more details.
+  /// @param l2OutputIndex The index of the output in the L2 block, or the dispute game index post fault proof upgrade.
+  /// @param outputRootProof The inclusion proof of of the L2ToL1MessagePasser contract's storage root.
```
Comment (matYang) [inline contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol]: nit: duplicate `of`

File: contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol [hunk 3]
```diff
@@ -5,33 +5,46 @@ import {IBridgeAdapter} from "../interfaces/IBridge.sol";
...
   // Nonce to use for L2 deposits to allow for better tracking offchain.
   uint64 private s_nonce = 0;
 
-  constructor(
-    IL1StandardBridge l1Bridge,
-    IWrappedNative wrappedNative,
-    IL1CrossDomainMessenger l1CrossDomainMessenger
-  ) {
-    if (address(l1Bridge) == address(0) || address(wrappedNative) == address(0)) {
+  constructor(IL1StandardBridge l1Bridge, IWrappedNative wrappedNative, IOptimismPortal optimismPortal) {
```
Comment (matYang) [inline contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol:77]: nit: typically constructor is defined after err/event/types

File: contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol [hunk 4]
```diff
@@ -72,41 +85,102 @@ contract OptimismL1BridgeAdapter is IBridgeAdapter {
...
+  /// @notice Optimism implements two-step withdrawals, so we need to specify the action to take
+  /// @notice each time the finalizeWithdrawERC20 function is called.
+  enum FinalizationAction {
+    Invalid,
+    ProveWithdrawal,
+    FinalizeWithdrawal
+  }
+
+  /// @notice used when the action in the payload is invalid.
+  error InvalidFinalizationAction();
```
Comment (matYang) [inline contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol]: nit: in CCIP error is defined at top, or we can stick with general Sol guide and make it below struct defs

File: contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol [hunk 5]
```diff
@@ -72,41 +85,102 @@ contract OptimismL1BridgeAdapter is IBridgeAdapter {
...
-      remoteSender,
-      localReceiver,
-      payload.amount,
-      data
-    );
+  /// @notice The action to take when finalizing a withdrawal.
+  /// @notice Optimism implements two-step withdrawals, so we need to specify the action to take
+  /// @notice each time the finalizeWithdrawERC20 function is called.
+  enum FinalizationAction {
+    Invalid,
```
Comment (matYang) [inline contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol]: `Invalid` shouldn't be needed here

File: contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol [hunk 6]
```diff
@@ -5,33 +5,46 @@ import {IBridgeAdapter} from "../interfaces/IBridge.sol";
...
 import {IL1StandardBridge} from "@eth-optimism/contracts/L1/messaging/IL1StandardBridge.sol";
-import {IL1CrossDomainMessenger} from "@eth-optimism/contracts/L1/messaging/IL1CrossDomainMessenger.sol";
 import {IERC20} from "../../vendor/openzeppelin-solidity/v4.8.3/contracts/token/ERC20/IERC20.sol";
 import {SafeERC20} from "../../vendor/openzeppelin-solidity/v4.8.3/contracts/token/ERC20/utils/SafeERC20.sol";
 
+import {Types} from "../interfaces/optimism/Types.sol";
+import {IOptimismPortal} from "../interfaces/optimism/IOptimismPortal.sol";
+
+/// @notice OptimismL1BridgeAdapter implements IBridgeAdapter for the Optimism L1<=>L2 bridge.
+/// @notice L1 -> L2:
```
Comment (matYang) [inline contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol]: nit: although there isn't a strict policy, I think typically we don't repeat @notice on every line; also consider keep notice short and use @dev for context details

File: contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL2BridgeAdapter.sol [hunk 7]
```diff
@@ -45,7 +72,12 @@ contract OptimismL2BridgeAdapter is IBridgeAdapter {
     // If the token is the wrapped native, we unwrap it and deposit native
     if (localToken == address(i_wrappedNative)) {
       i_wrappedNative.withdraw(amount);
-      i_L2Bridge.withdrawTo(Lib_PredeployAddresses.OVM_ETH, recipient, amount, 0, extraData);
+      // XXX: Lib_PredeployAddresses.OVM_ETH is actually 0xDeadDeAddeAddEAddeadDEaDDEAdDeaDDeAD0000.
```
Comment (matYang) [inline contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL2BridgeAdapter.sol:75]: LMAO nice address

File: contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL2BridgeAdapter.sol [hunk 8]
```diff
@@ -45,7 +72,12 @@ contract OptimismL2BridgeAdapter is IBridgeAdapter {
...
+      // This code path still works because the L2 bridge is hardcoded to handle this specific address.
+      // The better approach might be to use the bridgeEthTo function, which is on the StandardBridge
+      // abstract contract, inherited by both L1StandardBridge and L2StandardBridge.
+      // This is also marked as legacy, so it might mean that this will be deprecated soon.
```
Comment (matYang) [inline contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL2BridgeAdapter.sol:79]: curious why not use `bridgeEthTo` if it's better supported?

File: contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol [hunk 9]
```diff
@@ -72,41 +85,102 @@ contract OptimismL1BridgeAdapter is IBridgeAdapter {
...
+  /// @notice in the same transaction.
+  /// @param action The action to take; either ProveWithdrawal or FinalizeWithdrawal.
+  /// @param data The payload for the action. If ProveWithdrawal, it must be an abi-encoded OptimismProveWithdrawalPayload.
+  ///        If FinalizeWithdrawal, it must be an abi-encoded OptimismFinalizationPayload.
+  struct FinalizeWithdrawERC20Payload {
+    FinalizationAction action;
+    bytes data;
+  }
+
+  function finalizeWithdrawERC20(
```
Comment (matYang) [inline contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol:134]: let's add success and fail test cases for this in Foundry

File: core/scripts/ccip/rebalancer/opstack/finalize.go [hunk 10]
```diff
@@ -36,21 +42,39 @@ func FinalizeL1(
...
+				Data:     messagePassed.Data,
+			},
+		},
+	)
+	helpers.PanicErr(err)
+
+	// then encode the finalize withdraw erc20 payload next.
+	encodedPayload, err := encoderABI.Methods["encodeFinalizeWithdrawalERC20Payload"].Inputs.Pack(
+		optimism_l1_bridge_adapter_encoder.OptimismL1BridgeAdapterFinalizeWithdrawERC20Payload{
+			Action: 2, // 2 for finalize withdrawal action
```
Comment (matYang) [inline core/scripts/ccip/rebalancer/opstack/finalize.go]: let's make this a constant

File: core/scripts/ccip/rebalancer/opstack/finalize.go [hunk 11]
```diff
@@ -36,21 +42,39 @@ func FinalizeL1(
...
+			Data:   encodedFinalizeWithdrawal,
 		},
 	)
 	helpers.PanicErr(err)
 
+	tx, err := l1Adapter.FinalizeWithdrawERC20(
+		env.Transactors[l1ChainID],
+		common.HexToAddress("0x0"), // doesn't matter
```
Comment (matYang) [inline core/scripts/ccip/rebalancer/opstack/finalize.go]: nit: I think you can just do `common.Address{}`, or `utils.ZeroAddress`

File: contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol (hunk 6 above)
Comment (makramkd) [reply contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol]: I'm actually not sure how natspec would render this, I guess I can give it a shot. I've seen notice and dev repeated on every line, e.g in [VRF.sol](https://github.com/smartcontractkit/ccip/blob/f416505db7de218b0a94a40b1e7d1d11274ee335/contracts/src/v0.8/vrf/VRF.sol#L5).

File: contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL2BridgeAdapter.sol (hunk 8 above)
Comment (makramkd) [reply contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL2BridgeAdapter.sol:79]: I'm actually not sure what the difference is, it'd require a whole investigation. I should reach out to Optimism and ask them what's the dealio. I think changing this is better in another PR to keep this PR self-contained. The finalization process is the same either way.

File: contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol (hunk 6 above)
Comment (makramkd) [reply contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol]: Doesn't seem to be a big difference other than whitespace in the generated docs. Will remove repeats.

File: contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol (hunk 3 above)
Comment (makramkd) [reply contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol:77]: Moved stuff around so now this is the case.

File: contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol (hunk 9 above)
Comment (makramkd) [reply contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol:134]: Added.

File: contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL2BridgeAdapter.sol (hunk 8 above)
Comment (matYang) [reply contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL2BridgeAdapter.sol:79]: sounds good, please do confirm with them, OP team is kinda known for not being the best with compatibility

**Review by matYang — APPROVED** [review]

**Pull request merged**